File: Python/minimize-or-of-remaining-elements-using-operations.py
# Time:  O(nlogr)
# Space: O(1)

# bitmasks, greedy
class Solution(object):
    def minOrAfterOperations(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        result = 0
        l = max(nums).bit_length()
        mask = (1<<l)-1
        for i in reversed(xrange(l)):
            result <<= 1
            curr, cnt = mask, 0
            for x in nums:
                curr &= x>>i
                if curr&~result:
                    cnt += 1
                else:
                    curr = mask
            if cnt > k:
                result += 1
        return result

File: C++/minimize-or-of-remaining-elements-using-operations.cpp
// Time:  O(nlogr)
// Space: O(1)

// bitmasks, greedy
class Solution {
public:
    int minOrAfterOperations(vector<int>& nums, int k) {
        const auto& bit_length = [](int x) {
            return (x ? std::__lg(x) : -1) + 1;
        };

        int result = 0;
        const int l = bit_length(ranges::max(nums));
        const int mask = (1 << l) - 1;
        for (int i = l - 1; i >= 0; --i) {
            result <<= 1;
            int curr = mask, cnt = 0;
            for (const auto& x : nums) {
                curr &= x >> i;
                if (curr & ~result) {
                    ++cnt;
                } else {
                    curr = mask;
                }
            }
            if (cnt > k) {
                ++result;
            }
        }
        return result;
    }
};
